### PR TITLE
Feature/retrieve legal unit

### PR DESCRIPTION
**Pull request by urquha**
Added happy path for retrieving legal units using the local unit retrieval as a template, handled sad paths and technical issues.

### CODE REVIEW

File: app/repository/hbase/legalunit/LegalUnitColumns.scala [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package repository.hbase.legalunit
+
+object LegalUnitColumns {
+  val uBRN = "UBRN"
+  val UBRNref = "UBRNref"
+  val ern = "ern"
+  val entref = "entref"
+
```
Comment (ameenhaq) [inline app/repository/hbase/legalunit/LegalUnitColumns.scala]: Need to add the other columns even those that are optional (as declared in LegalUnitRowMapper.scala)

File: app/repository/hbase/legalunit/LegalUnitColumns.scala [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package repository.hbase.legalunit
+
+object LegalUnitColumns {
+  val uBRN = "UBRN"
```
Comment (ameenhaq) [inline app/repository/hbase/legalunit/LegalUnitColumns.scala]: maybe have uBRN as ubrn instead may look cleaner

File: test/uk/gov/ons/sbr/models/legalUnit/LegalUnitSpec.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        Json.toJson(SampleAllValuesLegalUnit) shouldBe Json.parse(expectedJsonStrOf(SampleAllValuesLegalUnit))
+      }
+
+      "when only the mandatory fields are defined" in new Fixture {
+        Json.toJson(SampleMandatoryValuesLegalUnit) shouldBe Json.parse(expectedJsonStrOf(SampleMandatoryValuesLegalUnit))
+      }
+    }
+
+    "A LocalUnit" - {
+      "can be represented as JSON" - {
```
Comment (ameenhaq) [inline test/uk/gov/ons/sbr/models/legalUnit/LegalUnitSpec.scala]: There is no LocalUnit in a Legal Unit. Local Unit has a relationship with Enterprise and LocalUnit

File: test/uk/gov/ons/sbr/models/legalUnit/LegalUnitSpec.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        "when all fields are defined" in new Fixture {
+          Json.toJson(SampleAllValuesLegalUnit) shouldBe Json.parse(expectedJsonStrOf(SampleAllValuesLegalUnit))
+        }
+
+        "when only the mandatory fields are defined" in new Fixture {
+          Json.toJson(SampleMandatoryValuesLegalUnit) shouldBe Json.parse(expectedJsonStrOf(SampleMandatoryValuesLegalUnit))
+        }
```
Comment (ameenhaq) [inline test/uk/gov/ons/sbr/models/legalUnit/LegalUnitSpec.scala]: Think this test is duplicated

File: app/repository/hbase/legalunit/LegalUnitRowMapper.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+/*
+ * The following fields are optional:
+ * - luref
+ * - entref
+ * - tradingstyle
+ * - address2
+ * - address3
+ * - address4
+ * - address5
+ *
```
Comment (ameenhaq) [inline app/repository/hbase/legalunit/LegalUnitRowMapper.scala]: These should be declared in your LegalUnitsColumns

File: app/uk/gov/ons/sbr/models/legalunit/LegalUnit.scala [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import io.swagger.annotations.ApiModelProperty
+import play.api.libs.json.{ Json, OWrites }
+
+import uk.gov.ons.sbr.models.enterprise.EnterpriseLink
+
+case class LegalUnit(
+  @ApiModelProperty(value = "Legal Unit Reference Number (UBRN)", dataType = "string", example = "1000000012", required = true) uBRN: UBRN,
+  @ApiModelProperty(value = "IDBR Local Unit Reference", dataType = "string", example = "9999999999", required = false) UBRNref: Option[String],
+  @ApiModelProperty(value = "A container for links to the associated Enterprise", dataType = "uk.gov.ons.sbr.models.enterprise.EnterpriseLink", required = true) enterprise: EnterpriseLink
+
```
Comment (ameenhaq) [inline app/uk/gov/ons/sbr/models/legalunit/LegalUnit.scala]: Need to update this to reflect the actual LegalUnit structure - there is no EnterpriseLink for example from what I know in LegalUnit

File: test/it/RetrieveAllLegalUnitsForEnterpriseAcceptanceSpec.scala [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+      When(s"the legal units with $TargetErn for $TargetPeriod are requested")
+      val response = await(wsClient.url(s"/v1/enterprises/${TargetErn.value}/periods/${Period.asString(TargetPeriod)}/legalunits").get())
+
+      Then("a NOT FOUND response is returned")
+      response.status shouldBe NOT_FOUND
+    }
+    scenario("when the specified period is invalid") { wsClient =>
+      Given(s"that a valid Period has yyyyMM format")
+      val invalidPeriod = "2018-02"
+      stubHBaseFor(anAllLegalUnitsForEnterpriseRequest(withErn = TargetErn, withPeriod = TargetPeriod).willReturn(
```
Comment (nigelhp) [inline test/it/RetrieveAllLegalUnitsForEnterpriseAcceptanceSpec.scala]: Do we need to stub here?  The router should reject the request, and so we should never make a REST call.

File: test/it/RetrieveLegalUnitBYKeyAcceptanceSpec.scala [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+import play.api.http.Status.{ BAD_REQUEST, NOT_FOUND, OK, INTERNAL_SERVER_ERROR }
+import play.mvc.Http.MimeTypes.JSON
+import repository.hbase.legalunit.LegalUnitColumns._
+import repository.hbase.legalunit.LegalUnitQuery
+import support.WithWireMockHBase
+import uk.gov.ons.sbr.models.Period
+import uk.gov.ons.sbr.models.enterprise.{ EnterpriseLink, Ern }
+import uk.gov.ons.sbr.models.legalunit.{ LegalUnit, UBRN }
+
+class RetrieveLegalUnitBYKeyAcceptanceSpec extends ServerAcceptanceSpec with WithWireMockHBase with OptionValues {
```
Comment (nigelhp) [inline test/it/RetrieveLegalUnitBYKeyAcceptanceSpec.scala]: capitalised Y in BY

File: test/it/RetrieveLegalUnitBYKeyAcceptanceSpec.scala [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      val response = await(wsClient.url(s"/v1/enterprises/123456789/periods/${Period.asString(TargetPeriod)}/legalunits/${TargetUBRN.value}").get())
+
+      Then(s"a BAD REQUEST response is returned")
+      response.status shouldBe BAD_REQUEST
+    }
+  }
+  feature("handle inability to connect to HBase REST service") {
+    scenario("WireMock stops, imitating a loss of connection to HBase") { wsClient =>
+      Given("request sent to the HBase REST mock")
+      stubHBaseFor(aLegalUnitRequest(withErn = TargetErn, withPeriod = TargetPeriod, withUBRN = TargetUBRN).willReturn(
```
Comment (nigelhp) [inline test/it/RetrieveLegalUnitBYKeyAcceptanceSpec.scala]: We do not need to stub here - as we are about to shutdown the server anyway.

File: test/it/RetrieveLegalUnitBYKeyAcceptanceSpec.scala [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        anOkResponse().withBody(LegalUnitSingleMatchHBaseResponseBody)
+      ))
+
+      When("the mock of HBase stops")
+      stopWireMock()
+      val response = await(wsClient.url(s"/v1/enterprises/${TargetErn.value}/periods/${Period.asString(TargetPeriod)}/legalunits/${TargetUBRN.value}").get())
+
+      Then("an INTERNAL SERVER ERROR response is returned")
+      response.status shouldBe INTERNAL_SERVER_ERROR
+    }
```
Comment (nigelhp) [inline test/it/RetrieveLegalUnitBYKeyAcceptanceSpec.scala]: This doesn't really tell the story from the user perspective.  Aim for something like:

Given("the legal unit repository is unavailable")
stopWireMock()

When("the user requests a legal unit")
val response = await(wsClient.url(...

Then("an INTERNAL SERVER ERROR response is returned")
response.status shouldBe INTERNAL_SERVER_ERROR

File: test/repository/hbase/legalunit/HBaseRestLegalUnitRepositorySpec.scala [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    val rowMapper: RowMapper[LegalUnit] = mock[RowMapper[LegalUnit]]
+    val config = HBaseRestLegalUnitRepositoryConfig(TargetTable)
+    val repository = new HBaseRestLegalUnitRepository(config, restRepository, rowMapper)
+
+    private val UnusedRowKey = ""
+    def toRow(variables: Map[String, String]) = Row(rowKey = UnusedRowKey, fields = variables)
+  }
+
+  private trait SingleResultFixture extends Fixture with SampleLegalUnit {
+    val TargetUBRN = UBRN("900000015")
```
Comment (nigelhp) [inline test/repository/hbase/legalunit/HBaseRestLegalUnitRepositorySpec.scala]: UBRN is not a realistic length?

File: test/repository/hbase/legalunit/HBaseRestLegalUnitRepositorySpec.scala [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+      "returning nothing when no legal units are found" in new MultipleResultFixture {
+        (restRepository.findRows _).expects(TargetTable, TargetQuery, DefaultColumnFamily).returning(Future.successful(Right(Seq.empty)))
+
+        whenReady(repository.findLegalUnitsForEnterprise(TargetErn, TargetPeriod)) { result =>
+          result.right.value shouldBe empty
+        }
+      }
+
+      "signalling failure when a valid Local Unit cannot be constructed from a successful HBase REST response" in new MultipleResultFixture {
```
Comment (nigelhp) [inline test/repository/hbase/legalunit/HBaseRestLegalUnitRepositorySpec.scala]: legal not local

File: test/uk/gov/ons/sbr/models/legalUnit/LegalUnitSpec.scala [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+import uk.gov.ons.sbr.models.legalunit.LegalUnit
+
+class LegalUnitSpec extends FreeSpec with Matchers {
+
+  private trait Fixture extends SampleLegalUnit {
+    def expectedJsonStrOf(legalUnit: LegalUnit): String =
+      s"""
+         |{${
+        withValues(
+          string("uBRN", legalUnit.uBRN.value),
```
Comment (nigelhp) [inline test/uk/gov/ons/sbr/models/legalUnit/LegalUnitSpec.scala]: In the Json representations, I believe our convention is lowercase - ubrn / ubrnref seem preferable.

File: app/repository/hbase/legalunit/LegalUnitColumns.scala (hunk 2 above)
Comment (awharris) [reply app/repository/hbase/legalunit/LegalUnitColumns.scala]: Agree, uBRN looks ugly, change to ubrn throughout the codebase

File: app/controllers/v1/LegalUnitController.scala [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  override def retrieveAllLegalUnitsForEnterprise(ernStr: String, periodStr: String): Action[AnyContent] = Action.async {
+    repository.findLegalUnitsForEnterprise(Ern(ernStr), Period.fromString(periodStr)).map { errorOrLocalUnits =>
+      errorOrLocalUnits.fold(resultOnFailure, resultOnSuccessWithMaybeManyUnits)
+    }
+  }
+  private def resultOnSuccessWithMaybeManyUnits(localUnits: Seq[LegalUnit]): Result =
+    if (localUnits.isEmpty) NotFound
+    else Ok(toJson(localUnits))
+
+  def badRequest(ernStr: String, periodStr: String, uBRNStrOpt: Option[String]) = Action {
```
Comment (awharris) [inline app/controllers/v1/LegalUnitController.scala]: ubrnStrOpt

File: app/repository/LegalUnitRepository.scala [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+
+import repository.RestRepository.ErrorMessage
+import uk.gov.ons.sbr.models.Period
+import uk.gov.ons.sbr.models.enterprise.Ern
+import uk.gov.ons.sbr.models.legalunit.{ LegalUnit, UBRN }
+
+import scala.concurrent.Future
+
+trait LegalUnitRepository {
+  def retrieveLegalUnit(ern: Ern, period: Period, uBRN: UBRN): Future[Either[ErrorMessage, Option[LegalUnit]]]
```
Comment (awharris) [inline app/repository/LegalUnitRepository.scala]: ubrn not uBRN

File: app/repository/hbase/legalunit/LegalUnitColumns.scala (hunk 1 above)
Comment (awharris) [reply app/repository/hbase/legalunit/LegalUnitColumns.scala]: There is no such thing as UBRNREF. Enterprises have two ids, the new id from SBR and the IDBR id. Legal units are new so only have UBRNs thus remove the UBRNref val.

File: app/repository/hbase/legalunit/LegalUnitRowMapper.scala (hunk 5 above)
Comment (awharris) [reply app/repository/hbase/legalunit/LegalUnitRowMapper.scala]: luref is local unit ref so remove from the comment

File: app/repository/hbase/legalunit/LegalUnitRowMapper.scala [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+ * Note that we use '=' for these fields within the body of the for expression rather than a generator '<-', so
+ * that we capture the field as an Option.
+ */
+object LegalUnitRowMapper extends RowMapper[LegalUnit] {
+
+  override def fromRow(variables: Row): Option[LegalUnit] =
+    for {
+      uBRN <- variables.fields.get(uBRN)
+      optUBRNref = variables.fields.get(UBRNref)
```
Comment (awharris) [inline app/repository/hbase/legalunit/LegalUnitRowMapper.scala]: As before there is no such thing as UBRNREF so remove

File: test/controllers/v1/LegalUnitControllerSpec.scala [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+import uk.gov.ons.sbr.models.legalunit.UBRN
+
+import scala.concurrent.Future
+
+class LegalUnitControllerSpec extends FreeSpec with Matchers with MockFactory with OptionValues {
+
+  private trait Fixture extends SampleLegalUnit {
+    val TargetErn = Ern("1234567890")
+    val TargetPeriod = Period.fromYearMonth(2018, FEBRUARY)
+    val TargetUBRN = UBRN("987654321")
```
[the remark this answers is not in the document]
Comment (awharris) [reply test/controllers/v1/LegalUnitControllerSpec.scala]: UBRNs are indeed 16 digits in length

File: app/repository/hbase/legalunit/LegalUnitColumns.scala (hunk 1 above)
Comment (awharris) [reply app/repository/hbase/legalunit/LegalUnitColumns.scala]: Address is being added to legal units (currently in BI there is only a postcode) so please include the address parts and postcode in the columns. Mandatory is address1 and postcode as per the other unit types.

**Pull request merged**